io/gdal/GDALReader.cpp

```diff
@@ -0,0 +1,216 @@
+//     for (auto di = dims.begin(); di != dims.end(); ++di)
+//         qi.m_dimNames.push_back(layout->dimName(*di));
+//     if (!Utils::numericCast(m_lasHeader.pointCount(), qi.m_pointCount))
+    qi.m_pointCount = m_raster->m_raster_x_size * m_raster->m_raster_y_size;
+//     qi.m_bounds = m_lasHeader.getBounds();
+    qi.m_srs = m_raster->getSpatialRef();
+    qi.m_valid = true;
+
+    PointTable table;
+    done(table);
+
+    return qi;
+}
+
+
+void GDALReader::addDimensions(PointLayoutPtr layout)
+{
+    int nBands = m_raster->m_band_count;
+
+    layout->registerDim(pdal::Dimension::Id::X);
+    layout->registerDim(pdal::Dimension::Id::Y);
+    for (int i=0; i < nBands; ++i)
+    {
+        std::ostringstream oss;
+        oss << "band-" << i+1;
+        layout->registerOrAssignDim(oss.str(), Dimension::Type::Double);
+    }
+}
+
+
+void GDALReader::ready(PointTableRef table)
+{
+    m_index = 0;
+
+
+}
+
+template<typename T>
+double convert(uint8_t*p)
+{
+    double output;
+
+    T t;
+    std::copy(p, p + sizeof(t), (uint8_t*)&t);
+    output = static_cast<double>(t);
+    return output;
+}
+
+point_count_t GDALReader::read(PointViewPtr view, point_count_t num)
+{
+    point_count_t count = std::min(count, m_count - m_index);
+    PointId nextId = view->size();
+
+    std::array<double, 2> coords;
+    for (int row = 0; row < m_raster->m_raster_y_size; ++row)
+    {
+        for(int col = 0; col < m_raster->m_raster_x_size; ++col)
+        {
+            m_raster->pixelToCoord(row, col, coords);
+            view->setField(Dimension::Id::X, nextId, coords[0]);
+            view->setField(Dimension::Id::Y, nextId, coords[1]);
+            nextId++;
+        }
+    }
+
+    std::vector<uint8_t> band;
+    std::vector<pdal::Dimension::Type::Enum> band_types = m_raster->getPDALDimensionTypes();
+    for (int b = 0; b < m_raster->m_band_count; ++b)
+    {
+        // Bands count from 1
+        // read up the band's bytes into an array,
+        // convert them from GDAL native types into
+        // doubles, and stuff them into the view
+        m_raster->readBand(band, b+1);
+        std::stringstream oss;
+        oss << "band-" << b+1;
+        log()->get(LogLevel::Info) << "Read band '" << oss.str() <<"'" << std::endl;
+        pdal::Dimension::Id::Enum d = view->layout()->findProprietaryDim(oss.str());
+        size_t dimSize = pdal::Dimension::size(band_types[b]); // count from 0
+
+        uint8_t* p = band.data();
+        for (point_count_t i = 0; i < count; ++i)
+        {
+            double test =  convert<uint8_t>(p);
+
+            if (band_types[b] == pdal::Dimension::Type::Float ||
+                band_types[b] == pdal::Dimension::Type::Double )
+            {
+                if (dimSize == 4)
+                    view->setField(d, i, convert<float>(p));
+                else
+                    view->setField(d, i, convert<double>(p));
+            }
+            else
+            {
+                if (dimSize == 1)
```

Comment:
If the data is properly aligned, you can just cast the pointer...
*(uint8_t *) p
*(uint16_t *) p
*(uint32_t *) p

PDAL will do the conversion to double for you if necessary.